Use optional type extension for Array3D fixed index slicing

nelson-wu opened the pull request:
This unbreaks https://github.com/dotnet/fsharp/issues/7735

TIHan commented:
@dsyme - whenever you get a chance if you can lookat it. The TLDR of this: Instead of adding the special cases in the type checker for array3d slicing, we will use the proper type extension.

I wonder what the reasoning was for adding the special cases in the type checker.

dsyme commented on src/fsharp/FSharp.Core/prim-types.fsi:
```diff
@@ -2814,63 +2814,6 @@ namespace Microsoft.FSharp.Core
-            /// <param name="start3">The start index of the third dimension.</param>
-            /// <param name="finish3">The end index of the third dimension.</param>
-            /// <returns>The two dimensional sub array from the given indices.</returns>
-            val inline GetArraySlice3DFixedSingle1 : source:'T[,,] ->  index1:int -> start2:int option -> finish2:int option -> start3:int option -> finish3:int option -> 'T[,]
```
These would have to remain in the DLL to avoid a binary breaking change for FSHarp.Core

dsyme commented on src/fsharp/FSharp.Core/prim-types.fsi:
```diff
@@ -3425,6 +3305,25 @@ namespace Microsoft.FSharp.Core
+    [<AutoOpen>]
+    module ArrayExtensions = 
+        type ``[,,]``<'T> with
+                member GetSlice : int * int option * int option * int option * int option -> 'T[,]
```
Need names on all the arguments please

dsyme replied:
Param help on each of the arguments please

dsyme commented on src/fsharp/TypeChecker.fs:
```diff
@@ -6356,24 +6356,12 @@ and TcIndexerThen cenv env overallTy mWholeExpr mDot tpenv wholeExpr e1 indexArg
                 | false, true, SynExpr.DotIndexedGet (_, [SynIndexerArg.Two _;SynIndexerArg.One _], _, _)                        -> Some (sliceOpPath, "GetArraySlice2DFixed2", GetIndexArgs indexArgs)
                 | false, true, SynExpr.DotIndexedGet (_, [SynIndexerArg.Two _;SynIndexerArg.Two _], _, _)                        -> Some (sliceOpPath, "GetArraySlice2D", GetIndexArgs indexArgs)
                 | false, true, SynExpr.DotIndexedGet (_, [SynIndexerArg.Two _;SynIndexerArg.Two _;SynIndexerArg.Two _], _, _)    -> Some (sliceOpPath, "GetArraySlice3D", GetIndexArgs indexArgs)
-                | false, true, SynExpr.DotIndexedGet (_, [SynIndexerArg.One _;SynIndexerArg.Two _;SynIndexerArg.Two _], _, _)    -> Some (sliceOpPath, "GetArraySlice3DFixedSingle1", GetIndexArgs indexArgs)
```
These should be conditional on a langversion test

cartermp replied:
In this case I don't think LangVersion is necessary. Although there is the potential for new behavior if someone shadows the implementation in FSharp.Core, I think this will be incredible rare and will likely yield the same behavior (since any different behavior would constitute a bug on behalf of the implementer of slices for these types).

dsyme commented on src/fsharp/FSharp.Core/prim-types.fs:
```diff
@@ -5690,6 +5677,39 @@ namespace Microsoft.FSharp.Core
+                SetArraySlice3DFixedSingle target source index3 start1 finish1 start2 finish2 0 1
+            
+            member target.SetSlice(index1: int, index2: int, start3: int option, finish3: int option, source: 'T[])
+                = SetArraySlice3DFixedDouble target source index1 index2 start3 finish3 2
```
`=` at the end of the line before, not the start of this line

nelson-wu commented:
> This looks ok - though feels like we should have some kind of test, and we must avoid the binary breaking change. It should also be under `/langversion`, and presumably that should be tested
> 
> Note that the contents of the quoted form of multi-dimensional array access will change. This should also be added as a test. I think it's an acceptable change.

Just a clarification -- the array3d fixed slicing code isn't shipped yet. They've been merged to the fsharp5 branch, but nobody has used it yet, so we haven't made any breaking changes.

cartermp commented:
I also don't think we can do any LangVersion switch here, since it's FSharp.Core. We'd just have to add APIs and not change any existing ones

dsyme commented:
@nelson-wu I' trying to work out the status of this "type extensions for slicing" w.r.t. the problem I mentioned regarding quotations.  

That is, run this program with F# 4.7 and `fsharp5`:
```fsharp
sprintf "%A" <@ fun (x: int[,]) -> x.[0..0,0..0] @>;;
```

Likewise for other affected slicing.  

With F# 4.7 the quotation contains a call to `GetArraySlice2D`.  

```
  "Lambda (x,
        Call (None, GetArraySlice2D,
              [x, NewUnionCase (Some, Value (0)), NewUnionCase (Some, Value (0)),
               NewUnionCase (Some, Value (0)), NewUnionCase (Some, Value (0))]))"
```

This needs to still be the case with `fsharp5` branch, otherwise it is a breaking change.  If it's not addressed yet my suggestion is to implement this via a reverse mapping back to the old entry points in `QuotationTranslator.fs`

nelson-wu commented:
> 
> 
> @nelson-wu I' trying to work out the status of this "type extensions for slicing" w.r.t. the problem I mentioned regarding quotations.
> 
> That is, run this program with F# 4.7 and `fsharp5`:
> 
> ```fsharp
> sprintf "%A" <@ fun (x: int[,]) -> x.[0..0,0..0] @>;;
> ```
> 
> Likewise for other affected slicing.
> 
> With F# 4.7 the quotation contains a call to `GetArraySlice2D`.
> 
> ```
>   "Lambda (x,
>         Call (None, GetArraySlice2D,
>               [x, NewUnionCase (Some, Value (0)), NewUnionCase (Some, Value (0)),
>                NewUnionCase (Some, Value (0)), NewUnionCase (Some, Value (0))]))"
> ```
> 
> This needs to still be the case with `fsharp5` branch, otherwise it is a breaking change. If it's not addressed yet my suggestion is to implement this via a reverse mapping back to the old entry points in `QuotationTranslator.fs`

@dsyme This PR does not impact any APIs that were present in F# 4.7, it only impacts **new** APIs that I added in [here](https://github.com/dotnet/fsharp/pull/7630). This PR does not refactor **every** slicing method into extensions, but **only new ones** that I added for F# 5.0.

In other words,

Expression | F# 4.7 | F# 5.0
------------| -------|--------
`<@ arr1d.[*] @>` |  compiles | not changed
`<@ arr2d.[*, *] @>` | compiles | not changed
`<@ arr2d.[1, *] @>` | compiles | not changed
`<@ arr3d.[*, *, *] @>` | compiles | not changed
`<@ arr3d.[1, *, *] @> ` | does not compile | **changed**
`<@ arr3d.[*, 1, *] @> ` | does not compile | **changed**
`<@ arr3d.[*, *, 1] @> ` | does not compile | **changed**
`<@ arr3d.[1, 1, *] @> ` | does not compile | **changed**
`<@ arr3d.[*, 1, 1] @> ` | does not compile | **changed**
`<@ arr3d.[1, *, 1] @> ` | does not compile | **changed**

TIHan commented:
This PR only changes what @nelson-wu added in the other PR for fsharp5: https://github.com/dotnet/fsharp/pull/7630 - therefore, there is no breaking change as we have not shipped fsharp5 officially.

But, what @dsyme is saying about code quotations is a bit worrisome. Meaning, whatever strategy we choose to represent slicing for Array3D/Array4D will now be permanent, even more so when new code quotations will expect the API; otherwise we would have to do a code quotation translation if we were to change it in the future.

At the moment, the current state of fsharp5's API for fixed index slicing on Array3D is consistent with Array1D and Array2D APIs - I'm referring to the intrinsic functions. This feels like the right move, but it breaks a test that expects to have custom slicing type extensions on Array3D and Array4D. Because of how slicing is handled for Arrays in TypeChecker.fs, it prevents custom slicing type extensions from being chosen/called. Our current thinking would be that we would just make the newly added fixed index slicing extensions actually be real type extensions so the breaking test would pass. This also feels right. But also feels a bit odd as it breaks consistency and expectations if one is writing a code quotation and seeing a different shape of APIs for Array3D/Array4D compared to Array1D/Array2D.

In regards to the quotation translator, I don't think we need to do anything there for this PR. If we want to make the rest of Array slicing APIs be type extensions, we would need to add additional code to translate the type extension method back into the intrinsic method call for code quotations to prevent a breaking change. That would be done in a separate PR and would go through the RFC process.

@dsyme I'm not sure myself what the right choice is here when adding new slicing APIs for arrays. I think we conceptually agree that we want fixed index slicing, but what should the public API actually be and what are we willing to give up? It feels unfortunate that there was syntax re-writing for slicing on arrays to begin with; it puts us into this pickle. I believe this was intended as the spec says "assume" on the extensions. They don't actually exist but it is assumed they are as the syntax re-writing takes care of it, but in a inconsistent way when compared to if it was an actual type extension. The inconsistency comes from the shape of the APIs and the in-ability for the slicing syntax sugar to pick up custom slicing type extensions that we added by the user outside of FSharp.Core.

cartermp commented:
The other thing to consider is that with F# 5.0 we can reserve the right to adjust the quoted form so that it's consistent.

nelson-wu commented:
I'm closing this because we've determined that we cannot eventually move all slicing methods to type extensions, because that would cause breaking changes when the new compiler is compiled with an old version of FSharp.Core. Thus, our alternative is to keep the state of fixed index slicing for 3D and 4D arrays as it is. This would mean that if a user had their own type extensions on Array3D/4D for fixed indexing, we would be breaking their code. Our workaround is to add a language version flag to disable the newly added Array3D/4D fixed index slicing methods, so your own extensions would still work.

dsyme commented:
@nelson-wu Could you give more details? It's not just the part about quotations is it?  I think that part is fixable...

TIHan commented:
IIRC, if we moved array slicing methods to be real type extensions and then reference an older version of FSharp.Core that does not have those type extensions, it would not compile as we would remove the special handling for arrays/strings in `TypeChecker.fs`. 

We would have to add an additional behavioral heuristic that would choose the type extension over the special handling for arrays/strings so long as the type extension really exists. It would effectively make the special handling for arrays/strings to be fallbacks and still exist in the codebase. To me, this adds more behavioral complexity to the design and it might be wise to follow the same spirit as the current design does.
